fog/ledger/server/src/db_fetcher.rs

```diff
@@ -0,0 +1,312 @@
+        );
+
+        Self {
+            join_handle,
+            stop_requested,
+            shared_state,
+            num_queued_records_limiter,
+        }
+    }
+
+    /// Stop and join the db poll thread
+    pub fn stop(&mut self) -> Result<(), ()> {
+        if let Some(join_handle) = self.join_handle.take() {
+            self.stop_requested.store(true, Ordering::SeqCst);
+            join_handle.join().map_err(|_| ())?;
+        }
+
+        Ok(())
+    }
+
+    /// Get a copy of the currently known missing block ranges.
+    /// This updates over time by the background worker thread.
+    pub fn known_missing_block_ranges(&self) -> Vec<BlockRange> {
+        self.shared_state().missing_block_ranges.clone()
+    }
+
+    /// Get the list of FetchedRecords that were obtained by the worker thread.
+    /// This also clears the queue so that more records could be fetched by
+    /// the worker thread. This updates over time by the background worker
+    /// thread.
+    pub fn get_pending_fetched_records(&self) -> Vec<FetchedRecords> {
+        // First grab all the records queued so far.
+        let records = self.shared_state().fetched_records.split_off(0);
+
+        // Now, signal the condition variable that the queue has been drained.
+        let (lock, condvar) = &*self.num_queued_records_limiter;
+        let mut num_queued_records = lock.lock().expect("mutex poisoned");
+        *num_queued_records = 0;
+
+        counters::DB_FETCHER_NUM_QUEUED_RECORDS.set(0);
+
+        condvar.notify_one();
+
+        // Return the records
+        records
+    }
+
+    /// Get a locked reference to the shared state.
+    fn shared_state(&self) -> MutexGuard<DbFetcherSharedState> {
+        self.shared_state.lock().expect("mutex poisoned")
+    }
+}
+
+impl Drop for DbFetcher {
+    fn drop(&mut self) {
+        let _ = self.stop();
+    }
+}
+
+struct DbFetcherThread<DB: Ledger, E: LedgerEnclaveProxy + Clone + Send + Sync + 'static> {
+    db: DB,
+    stop_requested: Arc<AtomicBool>,
+    shared_state: Arc<Mutex<DbFetcherSharedState>>,
+    block_tracker: BlockTracker,
+    num_queued_records_limiter: Arc<(Mutex<usize>, Condvar)>,
+    logger: Logger,
+    enclave: E,
+}
+
+/// Background worker thread implementation that takes care of periodically
+/// polling data out of the database. Add join handle
+impl<DB: Ledger, E: LedgerEnclaveProxy + Clone + Send + Sync + 'static> DbFetcherThread<DB, E> {
+    pub fn start(
+        db: DB,
+        stop_requested: Arc<AtomicBool>,
+        shared_state: Arc<Mutex<DbFetcherSharedState>>,
+        num_queued_records_limiter: Arc<(Mutex<usize>, Condvar)>,
+        logger: Logger,
+        enclave: E,
+    ) {
+        let thread = Self {
+            db,
+            stop_requested,
+            shared_state,
+            block_tracker: BlockTracker::new(logger.clone()),
+            num_queued_records_limiter,
+            logger,
+            enclave,
+        };
+        thread.run();
+    }
```

Comment:
This is not actually a background thread -- there is no actual second thread being created and no join handle, and whoever calls this function will be blocked forever. This needs to be fixed before we can merge, it should have a call to spawn a thread which calls run, and the object should contain a join handle, like the db fetcher in fog view. This is because there might need to be a different background thread added later to this server.

Reply:
I'm sorry, I understand now -- the join handle is in the caller. I feel like some of this stuff is awkwardly named, but this is the pattern we have in fog-view so it should be fine for now

Reply:
Was able to verify that code is already in dbfecter